multi_import/object_cache.py

```diff
@@ -21,70 +22,97 @@ def __hash__(self):
         return self.obj.__hash__()
 
 
-class ObjectCache(defaultdict):
+class ObjectCache(object):
     multiple_objects_error = MultipleObjectsReturned
 
     def __init__(self, lookup_fields):
-        super(ObjectCache, self).__init__(self._default_factory)
+        self.cache_fields = self._flatten_fields(lookup_fields)
         self.lookup_fields = lookup_fields
-        self.cached_object_count = 0
+        self.object_count = 0
+        self.objects = defaultdict(lambda: defaultdict(set))
 
-    def _default_factory(self):
-        return defaultdict(set)
-
-    def _add_lookup_value(self, field, value, instance):
-        if not value:
-            return
-
-        lookup_dict = self[field]
-        key = unicode(value)
-
-        lookup_dict[key].add(
-            CachedObject(instance)
+    def __len__(self):
+        return self.object_count
+
+    def add(self, obj):
+        cached_obj = CachedObject(obj)
+
+        fields_to_cache = (
+            (field, value)
+            for field, value in
+            (
+                (field, getattr(obj, field, None))
+                for field in self.cache_fields
+            )
+            if value is not None
         )
 
-    def _get_lookup_value(self, field, value):
-        if field not in self or value is None:
-            return None
+        for field, value in fields_to_cache:
+            self.objects[field][unicode(value)].add(cached_obj)
+
+        self.object_count += 1
 
-        lookup_dict = self[field]
-        key = unicode(value)
+    def get(self, field, value, default=None):
+        if isinstance(field, six.string_types):
+            field = (field,)
+            value = (value,)
 
-        instance_set = lookup_dict[key]
-        if not instance_set:
-            return None
+        zipped_values = list(zip(field, value))
 
-        if len(instance_set) > 1:
-            raise self.multiple_objects_error
+        for f, v in zipped_values:
+            if f not in self.objects or v is None:
+                return default
```

Comment:
Yes, after discussion, this IS the correct behaviour because its an AND that we are looking for.  If one of the values is none, than it really does not match.  Perhaps a comment here would be nice?